netmap/attributes.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func subnetAttributeKey(id *refs.SubnetID) string {
+	txt, _ := id.MarshalText() // never returns an error
+
+	return attrSubnetPrefix + string(txt)
+}
+
+// WriteSubnetInfo writes NodeSubnetInfo to NodeInfo via attributes. NodeInfo must not be nil.
+//
+// Does not add (removes existing) attribute if node:
+//   * exists non-zero subnet;
+//   * enters zero subnet.
+//
+// Attribute key is calculated from ID using format `__NEOFS__SUBNET_%s`.
+// Attribute Value is:
+//   * `True` if node enters the subnet;
+//   * `False`, otherwise.
+func WriteSubnetInfo(node *NodeInfo, info NodeSubnetInfo) {
+	attrs := node.GetAttributes()
+
+	id := info.ID()
+	enters := info.Enters()
+
+	// calculate attribute key
+	key := subnetAttributeKey(id)
+
+	if refs.IsZeroSubnet(id) == enters {
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs = append(attrs[:i], attrs[i+1:]...)
+			}
+		}
+	} else {
+		var val string
+
+		if enters {
+			val = attrSubnetValEntry
+		} else {
+			val = attrSubnetValExit
+		}
+
+		presented := false
+
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs[i].SetValue(val)
+				presented = true
+			}
+		}
+
+		if !presented {
+			var attr Attribute
+
+			attr.SetKey(key)
+			attr.SetValue(val)
+
+			attrs = append(attrs, &attr)
+		}
+	}
+
+	node.SetAttributes(attrs)
+}
+
+// ErrRemoveSubnet is returned when a node needs to leave the subnet.
+var ErrRemoveSubnet = errors.New("remove subnet")
+
+var errNoSubnets = errors.New("no subnets")
+
+// IterateSubnets iterates over all subnets the node belongs to and passes the IDs to f.
+// Handler must not be nil.
+//
+// Subnet attributes are expected to be key-unique, otherwise undefined behavior.
+//
+// If f returns ErrRemoveSubnet, then removes subnet entry. Note that this leads to an instant mutation of NodeInfo.
+// Breaks on any other non-nil error and returns it.
+//
+// Returns an error if any subnet attribute has wrong format.
+// Returns an error if the node is not included in any subnet by the end of the loop.
+func IterateSubnets(node *NodeInfo, f func(refs.SubnetID) error) error {
+	attrs := node.GetAttributes()
+
+	var (
+		err     error
+		id      refs.SubnetID
+		entries uint
+
+		zeroEntry = true
+	)
+
+	for i := 0; i < len(attrs); i++ { // range must not be used because of attrs mutation in body
+		key := attrs[i].GetKey()
+
+		// cut subnet ID string
+		idTxt := strings.TrimPrefix(key, attrSubnetPrefix)
+		if idTxt == key {
+			// not a subnet attribute
+			continue
+		}
+
+		// check value
+		val := attrs[i].GetValue()
+		if val != attrSubnetValExit && val != attrSubnetValEntry {
+			return fmt.Errorf("invalid attribute value: %s", val)
+		}
+
+		// decode subnet ID
+		if err = id.UnmarshalText([]byte(idTxt)); err != nil {
+			return fmt.Errorf("invalid ID text: %w", err)
+		}
+
+		// update status of zero subnet
+		isZero := refs.IsZeroSubnet(&id)
+
+		if isZero {
+			zeroEntry = val == attrSubnetValEntry
+			if zeroEntry {
```

Comment:
This `if` is redundant because no matter of `zeroEntry` value, you will continue the loop. Either there or down at `if val == attrSubnetValExit`.

Reply:
Removed the condition.